pkg/sources/generator/tickgen.go

```diff
@@ -389,18 +389,19 @@ func (mg *memgen) generator(ctx context.Context, rate int, timeunit time.Duratio
 }
 
 func (mg *memgen) newReadMessage(key string, payload []byte, offset int64) *isb.ReadMessage {
+	readOffset := util.NewSimpleIntPartitionOffset(offset, mg.vertexInstance.Replica)
```

Comment:
It seems when a source doesn't naturally support partitioning, we use the `vertexInstance.Replica` as the partition index. Is the Replica mapped to a pod in the vertex? I thought we are moving from using the pod as a processing unit to using the partition.

Or the Replica isn't really a partition index and we are just using it here to build the unique ID? If so, would it be better if we use the following statement?

```
readOffset := util.NewSimpleIntPartitionOffset(strconv.FormatInt(offset, 10) + "-" + strconv.FormatInt(int64(mg.vertexInstance.Replica), 10), 0) // 0 is the partition index
```

Reply:
we need either the source partition or the replica id. either of the ones is required for the correctness of the watermark.

Reply:
> we need either the source partition or the replica id.

Yes, I agree, but I thought this PR is to choose partition over replica explicitly.

Sorry, I have two questions:

1. Is this PR to move away from using the pod name to create a processing entity? If so, does it imply that the pod name is not needed in the read offset, hence Replica is not needed?

2. If we translate the `NewSimpleIntPartitionOffset` statement to native Golang initialization syntax, it will look like below:

```
&simpleIntPartitionOffset{
		seq:          offset,
		partitionIdx: mg.vertexInstance.Replica,
	}
```

Does `mg.vertexInstance.Replica` represent the partition index? e.g. for HTTP source, should partitionIdx always be 0?

Reply:
> Yes, I agree, but I thought this PR is to choose partition over replica explicitly.

This PR is to make sure we use the same "processing unit" we use for the first publish between the source and the source vertex, between the source vertex and the next vertex. Earlier, we were publishing the source vertex's offset timeline only based on pod even though the source supported a different sharding/partitioning scheme.


> Sorry, I have two questions:
> 
> 1. Is this PR to move away from using the pod name to create a processing entity? If so, does it imply that the pod name is not needed in the read offset, hence Replica is not needed?

Making sure the source vertex publishes the offset using the same sharding/partitioning semantics that the source provides.

> 2. If we translate the `NewSimpleIntPartitionOffset` statement to native Golang initialization syntax, it will look like below:
> 
> ```
> &simpleIntPartitionOffset{
> 		seq:          offset,
> 		partitionIdx: mg.vertexInstance.Replica,
> 	}
> ```
> 
> Does `mg.vertexInstance.Replica` represent the partition index? e.g. for HTTP source, should partitionIdx always be 0?

Yes.

Reply:
I see. Thanks for the clarification. Resolving this comment.